code/game/machinery/vending.dm

```diff
@@ -305,9 +305,14 @@
 			attack_hand(user)
 		return
 	else if(istype(W, /obj/item/weapon/coin) && premium.len > 0)
-		user.drop_item(src)
-		coin = W
-		user << "<span class='notice'>You insert [W] into [src].</span>"
+		if (isnull(coin))
+			user.drop_item()

```

Comment:
drop_hands(src) is the workaround for getting free premium items.
because it calls src.loc internally
so
instead getting inside of the vending machine
it goes to the floor
so pick up the coin get the premium item (because it's reference as coin in vending machine) holla!